From adfe086041d83f24d3ccb4c32335be41815ba445 Mon Sep 17 00:00:00 2001
From: Yufei Gu <yufei.gu@cloudera.com>
Date: Fri, 16 Feb 2018 14:09:50 -0800
Subject: [PATCH 2728/2748] YARN-3760. FSDataOutputStream leak in
 AggregatedLogFormat.LogWriter.close().
 Contributed by Haibo Chen.

Change-Id: I9195b59a075c9e085d66fdca734336f9e697edf4
(cherry picked from commit 0cacfe74d60ae51bcfe100e8d06e17eedc4aedf0)
---
 .../yarn/logaggregation/AggregatedLogFormat.java   |    9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-common/src/main/java/org/apache/hadoop/yarn/logaggregation/AggregatedLogFormat.java b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-common/src/main/java/org/apache/hadoop/yarn/logaggregation/AggregatedLogFormat.java
index 1ce7584..3cb5d3c 100644
--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-common/src/main/java/org/apache/hadoop/yarn/logaggregation/AggregatedLogFormat.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-common/src/main/java/org/apache/hadoop/yarn/logaggregation/AggregatedLogFormat.java
@@ -451,11 +451,14 @@ public void append(LogKey logKey, LogValue logValue) throws IOException {
 
     public void close() {
       try {
-        this.writer.close();
-      } catch (IOException e) {
+        if (writer != null) {
+          writer.close();
+        }
+      } catch (Exception e) {
         LOG.warn("Exception closing writer", e);
+      } finally {
+        IOUtils.closeStream(this.fsDataOStream);
       }
-      IOUtils.closeStream(fsDataOStream);
     }
   }
 
-- 
1.7.9.5

